### PR TITLE
password-generator extension from @joshuaiz

### PR DESCRIPTION
**Pull request by joshuaiz**
## Description

A simple password generator extension with options:
- number of characters (5–32)
- use numbers
- use special characters

When run, it generates the password and copies it to the clipboard.

## Type of change

- New extension

## Screencast

CleanShot 2022-01-27 at 18 27 19@2x

CleanShot 2022-01-27 at 18 27 33@2x

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x ] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

Thank you and thanks for the help on Slack!

### CODE REVIEW

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @joshuaiz 👋 

Thanks for your contribution 💪 

I have some feedback:

- Is there a reason 32 is the longest possible password?
- instead of showing a `<Detail/>` I would just show a toast, with style accordingly, like: `showToast(ToastStyle.Success, "Copied Password", generatedPassword);` which will make it faster to shuffle to a “new” one.
- You could add an action, so it's possible to Copy And Paste.
- With above, you could add a preference to set the default (Let me know if you need an example)
- Consider using a NPM package to allow more settings.

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by joshuaiz** [comment]
Thanks for this.

- reason 32 is the longest: many sites max out at this number but can certainly go higher e.g. 64
- Can definitely show a Toast. Just FYI this was my first attempt and the docs never explain clearly what a “Toast” is so it was a bit confusing at first. A simple 2-3 sentence explaining what a Toast is with pictures and what the options are in the docs would be super helpful for newbies like me.
- Can add copy and paste actions
- Yes can you provide an example of setting the default action?
- Not sure what you mean by using an NPM package to allow more settings - can you explain? 

Best,

Joshua

**Comment by pernielsentikaer** [comment]
> Thanks for this. - reason 32 is the longest: many sites max out at this number but can certainly go higher e.g. 64

I think that would better yes 😄 

> Can definitely show a Toast. Just FYI this was my first attempt and the docs never explain clearly what a “Toast” is so it was a bit confusing at first. A simple 2-3 sentence explaining what a Toast is with pictures and what the options are in the docs would be super helpful for newbies like me.

Check out this [example extension repo](https://github.com/raycast/extensions/tree/main/examples/api-examples) – in here you can test almost every component 💪

 > Yes can you provide an example of setting the default action?
 
It's done in the Twitch extension [with primaryaction preference].(https://github.com/raycast/extensions/pull/593/files)

> Not sure what you mean by using an NPM package to allow more settings - can you explain? 

I haven't checked if there are any cool packages at https://www.npmjs.com that could add more cool features in the long term, but I think it's good as it is 😄 💪

**Comment by joshuaiz** [comment]
Awesome - thanks so much for the feedback - super helpful. I’ll make some updates and then submit a new PR when ready.

And thanks for supporting the Recast community as you do - I know it is part of your job but you are doing a great job making it inclusive, helpful, and generous!

J

**Comment by pernielsentikaer** [comment]
> Awesome - thanks so much for the feedback - super helpful. I’ll make some updates and then submit a new PR when ready.

You don't need to make a new PR, you should just push to this PR (as you did when you fixed the linting problems) 💎 

> And thanks for supporting the Recast community as you do - I know it is part of your job but you are doing a great job making it inclusive, helpful, and generous!

Thanks a lot, It means a lot to me 😍 If you have any questions at all, catch me here or in [Slack](https://raycastcommunity.slack.com/) 😄 💯

**Comment by joshuaiz** [comment]
Hi @pernielsentikaer I think I'm almost ready to push again but I have one question. How can I capture the input of the text field for the number of characters using `Enter`? 

The only way I can find to do it is using the `SubmitFormAction` or `Action.SubmitForm` but both of those use Command-Enter which is not intuitive. Is there another way to capture this input?

**Comment by pernielsentikaer** [comment]
Hi @joshuaiz 👋 

The default behavior within Raycast would be `cmd+enter`, so I think we should keep it this way – looking forward for your push 💯

**Comment by mattisssa** [comment]
Hey @joshuaiz! What is the status of this PR? It seems like it's been here for a long time unfinished. Shall we probably close this one and once you're ready – open another PR?

**Comment by joshuaiz** [comment]
@mattisssa Hi! Sorry been busy with other projects but I'm ready to submit again...I'll push tonight.

**Comment by mattisssa** [comment]
No rush, just wanted to ensure that the PR is still alive 😉

File: extensions/password-generator/src/generate-password.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import { useState, Fragment } from "react";
+import {
+  ActionPanel,
+  Action,
+  Form,
+  SubmitFormAction,
```
Comment (pernielsentikaer) [inline extensions/password-generator/src/generate-password.tsx:6]: Deprecated, use [Action.SubmitForm](https://developers.raycast.com/api-reference/user-interface/actions#action.submitform) instead.

File: extensions/password-generator/src/generate-password.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  Detail,
+  copyTextToClipboard,
```
Comment (pernielsentikaer) [inline extensions/password-generator/src/generate-password.tsx:8]: Deprecated, use [Clipboard.copy](https://developers.raycast.com/api-reference/clipboard#copy) instead.

File: extensions/password-generator/src/generate-password.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  showToast,
+  ToastStyle,
```
Comment (pernielsentikaer) [inline extensions/password-generator/src/generate-password.tsx:10]: Deprecated, use [Toast.Style](https://developers.raycast.com/api-reference/feedback/toast#toast.style) instead.

File: extensions/password-generator/src/generate-password.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  const handleGeneratePassword = (values: any) => {
+    // console.log(values);
+
+    const length = values.lengthinput;
+    const lengthNumber = parseInt(length, 10);
+    // console.log("lengthNumber", lengthNumber);
+
+    const useNumbers = values.usenumbers === 1 ? true : false;
+    const useChars = values.usechars === 1 ? true : false;
```
Comment (pernielsentikaer) [inline extensions/password-generator/src/generate-password.tsx:30]:
```suggestion
    const useNumbers = values.usenumbers === 1;
    const useChars = values.usechars === 1;
```
Simplified a bit 😄

File: extensions/password-generator/src/generate-password.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      isError(true);
+      setPassword("Error");
+      if (lengthNumber < 5) {
+        errorMessage = "Password length must be greater than 4";
+      } else if (lengthNumber > 64) {
+        errorMessage = "Password length must be less than 65";
+      } else if (!Number.isFinite(lengthNumber)) {
+        errorMessage = "Password length must be a number";
+      }
+    }
```
Comment (pernielsentikaer) [inline extensions/password-generator/src/generate-password.tsx:49]: Most of this seems dead

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @joshuaiz 👋 

Thanks for taking your time to finish the extension 🔥 

I have some feedback and suggestions for you to address 💪  

- I'm still not happy about the `<Detail />` change, it feels unnecessary, our Bitwarden extension have password generator integrated, maybe you can get some ideas?

MqZwGRPdu93zGA8e

Let me know what you think 😄 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by sxn** [comment]
hi @pernielsentikaer @joshuaiz - I took the liberty to fix the deprecations and clean the extension up a bit. If you're fine with the changes, I think we can merge it 😸

**Comment by joshuaiz** [comment]
I'm all or it...what do I need to do next? Pull changes down then push up again?

**Review by pernielsentikaer — APPROVED** [review]
Approved 😄

**Pull request merged**